Patch/fix multiple order by

circulon opened the pull request:
This Fixes using multiple .order_by() statements
Also there is a select query syntax fix that may have been an issue when using group by/having and order by in the same query

circulon committed:
Fix possible issue where keyword would not be separated correctly

circulon committed:
Handle multiple order bys correctly

circulon committed:
Added a test for multiple order by statements

josephmancuso commented on src/masoniteorm/query/grammars/BaseGrammar.py:
```diff
@@ -253,11 +253,15 @@ def process_order_by(self):
+                sql += self.order_by_string().format(
+                    column=self._table_column_string(column), direction=direction.upper()
+                )
+            sql = "ORDER BY " + sql
```
Grammar classes should abstract all SQL. Although all of the databases we currently support are the same, Databases we may support in the future may not. it really should be calling the child grammar classes to get this (`self.order_by_string()`)

josephmancuso commented on src/masoniteorm/query/grammars/MSSQLGrammar.py:
```diff
@@ -131,7 +131,7 @@ def table_string(self):
         return "[{prefix}{table}]"
 
     def order_by_string(self):
-        return "ORDER BY {column} {direction}"
```
These need to go back

circulon replied:
Hmmm 
Yeah I understand the Grammar needs to be abstracted.
Unfortunately if the order_by_string is reverted  you end up with a query that looks something like this in the current 0.9.3 version

SELECT ..... ORDER BY "name" DESCORDER BY "age" ASC .....  

Which is obviously broken and why I moved the prefix in the first portion of the string.


That said to keep the Grammar abstracted how about something like this in the BaseGrammar class...  a class var 

order_by_prefix = "ORDER BY"

This would allow easy future changes in any child class while still being able to be used in the process_order_by() like so

sql = self.order_by_prefix + sql

WDYT?

josephmancuso replied:
Yeah I think that's fine. I don't really see what the difference is between prefixing it as a class property and putting it in these format string

We can do this: `sql = self.order_by_prefix + sql` and then maybe we can refactor later. I think the way I did some of the platform class coding is better for this situation anyway. Its a bit newer code and it handled this differently

circulon replied:
To see the difference use the test for multiple order by against the current 0.9 version and it will fail and you will see the reason why the format string will not work when it is used in this manner. 

Anyway I will update this PR in the morning.
Cheers

circulon replied:
@josephmancuso 
After thinking about this, looking at the existing format strings and keep the abstraction in place I have changed things slightly.

This adjustment now keeps formatting and keyword usage in their own class.
Let me know if there are  any pther considerations?

josephmancuso reviewed (changes requested):
Thanks for fixing! FYI all grammar classes should abstract SQL so we really can't have something like an `"ORDER BY"` in the `BaseGrammar` class. That text should really exist in the individual child grammar classes

circulon committed:
Separate criteria building from keyword context

Use order_by_format to build criteria and order_by_string for keyword inclusion with criteria

circulon committed:
Fixed Linting issue

josephmancuso commented on tests/postgres/grammar/test_select_grammar.py:
```diff
@@ -69,6 +69,12 @@ def can_compile_with_order_by(self):
+        """
+        self.builder.select('username').order_by('age', 'desc').order_by('name').to_sql()
+        """
+        return """SELECT "users"."username" FROM "users" ORDER BY "users"."age" DESC, "users"."name" ASC"""
```
So this is a little tricky but how the select grammars tests work (and really how all test classes will work in the future when things aren't as fluid as they have been) Is that there is a `BaseTestCaseSelectGrammar` (see line 8) that has the test cases. those are inherited to the child test classes and call the related test methods to make the right assertions per grammar.

So for example a test case inside the `BaseTestCaseSelectGrammar` has a test case like `test_can_compile_with_multiple_order_by` which calls this method here (`can_compile_with_multiple_order_by` without the `test_` part). This allows the abstraction of tests that are the same to trickle down to the child test classes. So adding a test in 1 place basically calls tests for every grammar. Keeping it DRY

What I'm trying to say is that this test is not actually running without the `test_can_compile_with_multiple_order_by` inside the `BaseTestCaseSelectGrammar`

circulon committed:
Fixed test not being called

circulon committed:
Added missing tests

circulon commented:
@josephmancuso 
Thanks for that explanation.

I havent done unit tests in python before so missed that kick off point.

Tests added for all grammars and now passing

josephmancuso reviewed (approved):
Looks good to me! 👍

josephmancuso commented:
Thanks!
